### ISSUE TITLE
Weird behavior of the DistributedOptimizer. 

### ISSUE THREAD

**Issue by ysyyork** [horovod/horovod#2579]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) Tensorflow 
2. Framework version: 2.3
3. Horovod version: 0.19.5
4. MPI version: 4.0.0
5. CUDA version: 11.1
6. NCCL version: 
7. Python version: 3.6
8. Spark / PySpark version:
9. OS and version: Ubuntu
10. GCC version:
11. CMake version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Bug report:**
Please describe erroneous behavior you're observing and steps to reproduce it.

I'm training a keras mobilenet classification model. When I train it with single GPU, it can converge to very high (near 99%) accuracy. However when I use multiple workers, the accuracy dropped to only 50% ish with 2 workers and 40% ish with 3 workers. The more workers I used the lower the accuracy I got. This is really weird. 

Here is the training code:
```
def main(cmd_args):
    if cmd_args.gpus != 'all':
        os.environ['CUDA_VISIBLE_DEVICES'] = cmd_args.gpus
    hvd.init()
    root_rank = cmd_args.root_rank

    gpus = tf.config.experimental.list_physical_devices('GPU')
    for gpu in gpus:
        # tf.config.experimental.set_memory_growth(gpu, cmd_args.enable_gpu_memory_auto_growth)
        tf.config.experimental.set_memory_growth(gpu, True)
    if gpus:
        tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], 'GPU')

    weights = None
    if cmd_args.weights is not None:
        weights = normpath(pathlib.Path(cmd_args.weights).expanduser().absolute())
    data_path = normpath(pathlib.Path(cmd_args.data_path).expanduser().absolute())

    verbose = 1 if hvd.rank() == root_rank else 0
    session_id = time.time()
    prefix = './checkpoints/{timestamp}_{session_id}_l2_{l2_rate}_lr_{lr}_batch_{batch_size}_folder_{folder}_{backbone}'.format(
        timestamp=datetime.now().strftime("%Y-%M-%d-%H-%m-%S-%f"),
        session_id=session_id,
        l2_rate=cmd_args.l2_rate,
        batch_size=cmd_args.batch_size,
        folder=data_path.split('/')[-1],
        lr=cmd_args.lr,
        backbone=cmd_args.backbone,
    )
    if cmd_args.weights_only:
        checkpoint_path = '{prefix}/cp.ckpt'.format(
            prefix=prefix
        )
    else:
        checkpoint_path = '{prefix}/model'.format(
            prefix=prefix
        )



    data_loader = DataLoader(
        path=data_path,
        batch_size=cmd_args.batch_size,
        workers=cmd_args.workers,
        max_queue_size=cmd_args.max_queue_size,
        val_split=cmd_args.val_split,
        shuffle=True,
        verbose=verbose
    )

    model = ModelWrapper(
        num_of_classes=data_loader.num_of_cls,
        l2_rate=cmd_args.l2_rate,
        weights=weights,
        backbone=cmd_args.backbone,
        verbose=verbose
    ).model # this is just a mobilenet model. 

    lr = cmd_args.lr * hvd.size() * 1
    if cmd_args.optimizer == 'adam':
        opt = Adam(lr)
    elif cmd_args.optimizer == 'sgd':
        opt = SGD(lr, momentum=0.9)
    opt = hvd.DistributedOptimizer(opt)

    from tensorflow.keras.losses import categorical_crossentropy, sparse_categorical_crossentropy
    from tensorflow.keras.metrics import binary_accuracy, categorical_accuracy, sparse_categorical_accuracy

    def custom_accuracy(y_true, y_pred):
        pred_index = tf.argmax(y_pred, axis=-1)
        gt_index = tf.argmax(y_true, axis=-1)
        matches = tf.cast(tf.math.count_nonzero(tf.cast(pred_index == gt_index, dtype=tf.float32)), tf.float32)
        total = tf.cast(tf.shape(y_true)[0], tf.float32)
        res = matches / total
        # tf.print(pred_index, summarize=30)
        # tf.print(gt_index, summarize=30)
        # tf.print(matches)
        # tf.print(total)
        # tf.print(res)
        # tf.print("----------------", summarize=30)


        return res

    model.compile(
        optimizer=opt,
        loss=categorical_crossentropy,
        metrics=[
            custom_accuracy, # was using the categorical_accuracy from keras but same result. 
            categorical_crossentropy,
        ],
    )


    prefetch_num = 32
    train_seq = data_loader.get_data_seq(mode='train')
    val_seq = data_loader.get_data_seq(mode='val')

    train_dataset = tf.data.Dataset.from_generator(
        train_seq.get_gen,
        output_types=(tf.float32, tf.float32)
    ).prefetch(prefetch_num)
    val_dataset = tf.data.Dataset.from_generator(
        val_seq.get_gen,
        output_types=(tf.float32, tf.float32)
    ).prefetch(prefetch_num)
    train_steps = len(train_seq) // hvd.size()
    val_steps = len(val_seq) // hvd.size()


    early_stop = EarlyStopping(patience=cmd_args.early_stop_patience)
    reduce_lr = ReduceLROnPlateau(factor=0.5, patience=cmd_args.reduce_lr_patience)





    def logging_on_epoch_begin(epoch, logs):
        if verbose:
            print("Session: {session_id}, Learning rate: {lr:e}".format(lr=model.optimizer.lr.numpy(),
                                                                        session_id=session_id))

    logging_callback = LambdaCallback(
        on_epoch_begin=logging_on_epoch_begin
    )
    final_lr = tf.convert_to_tensor(lr, dtype=tf.float32)
    warmup_steps = cmd_args.warmup_steps


    callbacks = [
        hvd.callbacks.BroadcastGlobalVariablesCallback(root_rank),
        hvd.callbacks.MetricAverageCallback(),
        hvd.callbacks.LearningRateWarmupCallback(
            warmup_epochs=warmup_steps,
            initial_lr=lr,
            verbose=verbose,
            steps_per_epoch=train_steps
        ),
        reduce_lr,
        early_stop,
        logging_callback,
    ]

    if not cmd_args.disable_checkpoint and hvd.rank() == root_rank:
        checkpoint = CustomKerasModelCheckpoint(
            checkpoint_path,
            monitor='val_loss',
            # monitor='loss',
            save_best_only=(not cmd_args.disable_save_best),
            save_weights_only=cmd_args.weights_only,
            mode='min',
        )
        callbacks.append(checkpoint)

    model.fit(
        train_dataset,
        steps_per_epoch=train_steps,
        validation_data=val_dataset,
        validation_steps=val_steps,
        epochs=cmd_args.epochs,
        initial_epoch=cmd_args.initial_epoch,
        callbacks=callbacks,
        verbose=verbose,
    )

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--data-path',
        required=True
    )
    parser.add_argument(
        '--lr',
        type=float,
        default=1e-4,
        help="""
        Learning rate, default 1e-4
        """
    )
    parser.add_argument(
        '--batch-size',
        type=int,
        default=32,
        help="""
        Batch size, default 32
        """
    )
    parser.add_argument(
        '--disable-checkpoint',
        action='store_true',
        help="""
        If checkpoint should be enabled.
        """
    )
    parser.add_argument(
        '--disable-save-best',
        action='store_true',
        help="""
        Should the save best be disabled
        """
    )
    parser.add_argument(
        '--weights',
        type=str,
        default=None,
        help="""
        The weights location to load 
        """,
    )
    parser.add_argument(
        '--weights-only',
        action='store_true',
        help="""
        If set, will only save weights otherwise save as a folder 
        """,
    )
    parser.add_argument(
        '--val-split',
        type=float,
        default=0.15,
        help="""
        Validation set split for simple data loader 
        """,
    )
    parser.add_argument(
        '--workers',
        type=int,
        default=10,
        help="""
        workers number of the data loader
        """,
    )
    parser.add_argument(
        '--max-queue-size',
        type=int,
        default=30,
        help="""
        max queue size of the data loader
        """,
    )
    parser.add_argument(
        '--initial-epoch',
        type=int,
        default=0,
        help="""
        Initial epoch to start with
        """,
    )
    parser.add_argument(
        '--epochs',
        type=int,
        default=300,
        help="""
            Total epochs to run
            """,
    )
    parser.add_argument(
        '--early-stop-patience',
        type=int,
        default=40,
        help="""
        Early stop patience
        """
    )
    parser.add_argument(
        '--reduce-lr-patience',
        type=int,
        default=15,
        help="""
        Reduce lr patience
        """
    )
    parser.add_argument(
        '--l2-rate',
        type=float,
        default=0.0001,
        help="""
        When training triplet model, you can sepecify the weight of triplet loss. 
        """
    )
    parser.add_argument(
        '--root-rank',
        default=0,
        type=int,
        help="""
            The root rank for horovod
            """
    )
    parser.add_argument(
        '-egmag', '--enable-gpu-memory-auto-growth',
        action='store_true',
        help='''
            Should gpu be in auto growth mode
            '''
    )
    parser.add_argument(
        '--optimizer',
        default='adam',
        choices=['adam', 'sgd']
    )
    parser.add_argument(
        '--warmup-steps',
        type=float,
        default=5
    )
    parser.add_argument(
        '--backbone',
        choices=['mobilenet_v2', 'mobilenet'],
        default='mobilenet_v2'
    )
    parser.add_argument(
        '--gpus',
        type=str,
        default='all'
    )
    cmd_args, unparsed = parser.parse_known_args()
    print(cmd_args)

    main(cmd_args)
```
However, if you comment out the line of `opt = hvd.DistributedOptimizer(opt)` and run with multiple workers, it turned normal. So I guess the issue is with this `DistributedOptimizer`. I also upgraded the horovod version to the latest 0.21.1 and change it to `opt = hvd.DistributedOptimizer(opt,  backward_passes_per_step=1, average_aggregated_gradients=True)` Problem still persist. Not sure what's happening here.

**Comment by ysyyork** [comment]
More info. 

Tested the same environment with a different model which is also a classification model. Dataset for that model is much bigger. And tested with single GPU as well as multi GPUs. results are similar. Maybe it's dataset issue since the dataset above are only 3k. But I didn't get it cus technically that shouldn't make any difference for single GPU and multi-gpu. Someone has any ideas about the reason?

**Comment by ysyyork** [comment]
Dumb me, it's my data issue. data loading on different servers is different which causes wrong labels.
